diff --git a/src/pcm.c b/src/pcm.c
index 5fcde93..203b738 100644
--- a/src/pcm.c
+++ b/src/pcm.c
@@ -742,7 +742,7 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 
     x.buf = data;
     x.frames = frame_count;
-    x.result = 0;
+
     for (;;) {
         if ((!pcm->running) && (pcm_start(pcm) < 0))
             return -errno;
@@ -756,7 +756,7 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
             }
             return oops(pcm, errno, "cannot read stream data");
         }
-        return x.result;
+        return 0;
     }
 }
 
