diff --git a/gst-libs/gst/audio/streamvolume.c b/gst-libs/gst/audio/streamvolume.c
index 9b7bf82..7be2df4 100644
--- a/gst-libs/gst/audio/streamvolume.c
+++ b/gst-libs/gst/audio/streamvolume.c
@@ -52,6 +52,8 @@
 #include "streamvolume.h"
 #include <math.h>

+#define VOLUME_STEPS 20
+
 static void
 gst_stream_volume_class_init (GstStreamVolumeInterface * iface)
 {
@@ -213,3 +215,17 @@ gst_stream_volume_convert_volume (GstStreamVolumeFormat from,
   }
   g_return_val_if_reached (0.0);
 }
+
+void gst_set_relative_volume (GstElement * play, gdouble volume_step)
+{
+   //gdouble volume;
+   //volume = gst_stream_volume_get_volume (GST_STREAM_VOLUME (play),
+   //GST_STREAM_VOLUME_FORMAT_CUBIC);
+
+   //volume = round ((volume + volume_step) * VOLUME_STEPS) / VOLUME_STEPS;
+   //volume = CLAMP (volume, 0.0, 10.0);
+
+   gst_stream_volume_set_volume (GST_STREAM_VOLUME (play),
+   GST_STREAM_VOLUME_FORMAT_CUBIC, volume_step/100);
+}
+
diff --git a/gst-libs/gst/audio/streamvolume.h b/gst-libs/gst/audio/streamvolume.h
index c8d16e9..ee9ee8a 100644
--- a/gst-libs/gst/audio/streamvolume.h
+++ b/gst-libs/gst/audio/streamvolume.h
@@ -72,6 +72,10 @@ gboolean        gst_stream_volume_get_mute        (GstStreamVolume *volume);
 gdouble         gst_stream_volume_convert_volume  (GstStreamVolumeFormat from,
                                                    GstStreamVolumeFormat to,
                                                    gdouble val) G_GNUC_CONST;
+
+GST_AUDIO_API
+void            gst_set_relative_volume            (GstElement * play,
+                                                   gdouble volume_step);

 G_END_DECLS

