diff --git a/tools/gst-play.c b/tools/gst-play.c
index 189e8f7..23fb995 100644
--- a/tools/gst-play.c
+++ b/tools/gst-play.c
@@ -102,6 +102,7 @@ static gboolean play_timeout (gpointer user_data);
 static void play_about_to_finish (GstElement * playbin, gpointer user_data);
 static void play_reset (GstPlay * play);
 static void play_set_relative_volume (GstPlay * play, gdouble volume_step);
+static void play_info_print (GstPlay * play);
 static gboolean play_do_seek (GstPlay * play, gint64 pos, gdouble rate,
     GstPlayTrickMode mode);

@@ -236,6 +237,9 @@ play_free (GstPlay * play)

   play_reset (play);

+  play->desired_state = GST_STATE_NULL;
+  play_info_print(play);
+
   gst_element_set_state (play->playbin, GST_STATE_NULL);
   gst_object_unref (play->playbin);

@@ -508,6 +512,8 @@ play_timeout (gpointer user_data)
     status[len] = '\0';
   }

+  play_info_print(play);
+
   if (pos >= 0 && dur > 0) {
     gchar dstr[32], pstr[32];

@@ -522,6 +528,16 @@ play_timeout (gpointer user_data)
   return TRUE;
 }

+static void play_info_print (GstPlay * play) {
+  gint64 pos = -1, dur = -1;
+  gchar *state[5] = {"GST_STATE_VOID_PENDING", "GST_STATE_NULL", "GST_STATE_READY", "GST_STATE_PAUSED", "GST_STATE_PLAYING"};
+
+  gst_element_query_position (play->playbin, GST_FORMAT_TIME, &pos);
+  gst_element_query_duration (play->playbin, GST_FORMAT_TIME, &dur);
+
+  GST_INFO ("[update_state] state=%s curtime=%lld (ms:%lld) fulltime=%lld\n", state[play->desired_state], pos / 1000000000, pos / 1000000, dur / 1000000000);
+}
+
 static gchar *
 play_uri_get_display_name (GstPlay * play, const gchar * uri)
 {
