Index: weston-2.99.91/clients/desktop-shell.c
===================================================================
--- weston-2.99.91.orig/clients/desktop-shell.c
+++ weston-2.99.91/clients/desktop-shell.c
@@ -1101,7 +1101,7 @@ background_create(struct desktop *deskto
 	background->widget = window_add_widget(background->window, background);
 	window_set_user_data(background->window, background);
 	widget_set_redraw_handler(background->widget, background_draw);
-	widget_set_transparent(background->widget, 0);
+	widget_set_transparent(background->widget, 1);
 
 	s = weston_config_get_section(desktop->config, "shell", NULL, NULL);
 	weston_config_section_get_string(s, "background-image",
