--- swupdate-2018.11/corelib/mtd-interface.c	2019-07-09 14:51:18.043766200 +0800
+++ swupdate-2018.11/corelib/mtd-interface.c	2019-07-09 14:32:41.761602400 +0800
@@ -83,7 +83,7 @@
 		}
 
 		/* Unlock memory if required */
-		if (mtd_is_locked(mtd, fd, eb)) {
+		if (mtd_is_locked(mtd, fd, eb) > 0) {
 			if (mtd_unlock(mtd, fd, eb) != 0) {
 				if (errno != EOPNOTSUPP) {
 					TRACE("%s: MTD unlock failure", mtd_device);
