diff --git a/examples/multimediawidgets/player/main.cpp b/examples/multimediawidgets/player/main.cpp
index a3dabe1..7dda031 100644
--- a/examples/multimediawidgets/player/main.cpp
+++ b/examples/multimediawidgets/player/main.cpp
@@ -44,6 +44,8 @@
 #include <QCommandLineParser>
 #include <QCommandLineOption>
 #include <QDir>
+#include <QGraphicsView>
+#include <QGraphicsProxyWidget>

 int main(int argc, char *argv[])
 {
@@ -67,11 +69,21 @@ int main(int argc, char *argv[])
             urls.append(QUrl::fromUserInput(a, QDir::currentPath(), QUrl::AssumeLocalFile));
         player.addToPlaylist(urls);
     }
-
+
+	player.resize(768,568);
+
 #if defined(Q_WS_SIMULATOR)
     player.setAttribute(Qt::WA_LockLandscapeOrientation);
     player.showMaximized();
 #else
+
+	//QGraphicsScene *scene = new QGraphicsScene;
+	//QGraphicsProxyWidget *w1 = scene->addWidget(&player);
+	//w1->setRotation(270);
+	//QGraphicsView *view = new QGraphicsView(scene);
+	//view->resize(768, 1024);
+	//view->show();
+
     player.show();
 #endif
     return app.exec();
