diff --git a/src/chrome/test/BUILD.gn b/src/chrome/test/BUILD.gn
index 96a581f..95eb911 100644
--- a/src/chrome/test/BUILD.gn
+++ b/src/chrome/test/BUILD.gn
@@ -3765,11 +3765,11 @@ test("unit_tests") {
       "../browser/ui/input_method/input_method_engine_unittest.cc",
     ]
   }
-  if (!is_chromeos && !is_chromecast && is_linux) {
-    sources +=
-        [ "../browser/ui/libgtkui/select_file_dialog_impl_gtk_unittest.cc" ]
-    deps += [ "//build/config/linux/gtk" ]
-  }
+#  if (!is_chromeos && !is_chromecast && is_linux) {
+#    sources +=
+#        [ "../browser/ui/libgtkui/select_file_dialog_impl_gtk_unittest.cc" ]
+#    deps += [ "//build/config/linux/gtk" ]
+#  }
 
   if (is_android || is_chromeos) {
     sources += [ "../browser/media/protected_media_identifier_permission_context_unittest.cc" ]
